// tsparticles/esm/Options/Classes/Interactivity/Modes/Trail.js
import _classCallCheck from "F:\\02-React\\05-Approve\\01-trydo\\trydo\\node_modules\\babel-preset-react-app\\node_modules\\@babel\\runtime/helpers/esm/classCallCheck";
import _createClass from "F:\\02-React\\05-Approve\\01-trydo\\trydo\\node_modules\\babel-preset-react-app\\node_modules\\@babel\\runtime/helpers/esm/createClass";
import {
    deepExtend
} from "../../../../Utils/Utils";
export var Trail = /*#__PURE__*/ function() {
    function Trail() {
        _classCallCheck(this, Trail);

        this.delay = 1;
        this.pauseOnStop = false;
        this.quantity = 1;
    }

    _createClass(Trail, [{
        key: "load",
        value: function load(data) {
            if (data === undefined) {
                return;
            }

            if (data.delay !== undefined) {
                this.delay = data.delay;
            }

            if (data.quantity !== undefined) {
                this.quantity = data.quantity;
            }

            if (data.particles !== undefined) {
                this.particles = deepExtend({}, data.particles);
            }

            if (data.pauseOnStop !== undefined) {
                this.pauseOnStop = data.pauseOnStop;
            }
        }
    }]);

    return Trail;
}();
